File: packages/rslint-test-tools/tests/eslint-plugin-jest/rules/no-restricted-matchers.test.ts
import { RuleTester } from '../rule-tester';

const ruleTester = new RuleTester();

ruleTester.run('no-restricted-matchers', {} as never, {
  valid: [
    { code: 'expect(a).toHaveBeenCalled()' },
    { code: 'expect(a).not.toHaveBeenCalled()' },
    { code: 'expect(a).toHaveBeenCalledTimes()' },
    { code: 'expect(a).toHaveBeenCalledWith()' },
    { code: 'expect(a).toHaveBeenLastCalledWith()' },
    { code: 'expect(a).toHaveBeenNthCalledWith()' },
    { code: 'expect(a).toHaveReturned()' },
    { code: 'expect(a).toHaveReturnedTimes()' },
    { code: 'expect(a).toHaveReturnedWith()' },
    { code: 'expect(a).toHaveLastReturnedWith()' },
    { code: 'expect(a).toHaveNthReturnedWith()' },
    { code: 'expect(a).toThrow()' },
    { code: 'expect(a).rejects;' },
    { code: 'expect(a);' },
    {
      code: 'expect(a).resolves',
      options: [{ not: null }],
    },
    {
      code: 'expect(a).toBe(b)',
      options: [{ 'not.toBe': null }],
    },
    {
      code: 'expect(a).toBeUndefined(b)',
      options: [{ toBe: null }],
    },
    {
      code: 'expect(a)["toBe"](b)',
      options: [{ 'not.toBe': null }],
    },
    {
      code: 'expect(a).resolves.not.toBe(b)',
      options: [{ not: null }],
    },
    {
      code: 'expect(a).resolves.not.toBe(b)',
      options: [{ 'not.toBe': null }],
    },
    {
      code: "expect(uploadFileMock).resolves.toHaveBeenCalledWith('file.name')",
      options: [
        { 'not.toHaveBeenCalledWith': 'Use not.toHaveBeenCalled instead' },
      ],
    },
    {
      code: "expect(uploadFileMock).resolves.not.toHaveBeenCalledWith('file.name')",
      options: [
        { 'not.toHaveBeenCalledWith': 'Use not.toHaveBeenCalled instead' },
      ],
    },
  ],
  invalid: [
    {
      code: 'expect(a).toBe(b)',
      options: [{ toBe: null }],
      errors: [
        {
          messageId: 'restrictedChain',
          data: {
            message: null,
            restriction: 'toBe',
          },
          column: 11,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(a)["toBe"](b)',
      options: [{ toBe: null }],
      errors: [
        {
          messageId: 'restrictedChain',
          data: {
            message: null,
            restriction: 'toBe',
          },
          column: 11,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(a).not[x]()',
      options: [{ not: null }],
      errors: [
        {
          messageId: 'restrictedChain',
          data: {
            message: null,
            restriction: 'not',
          },
          column: 11,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(a).not.toBe(b)',
      options: [{ not: null }],
      errors: [
        {
          messageId: 'restrictedChain',
          data: {
            message: null,
            restriction: 'not',
          },
          column: 11,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(a).resolves.toBe(b)',
      options: [{ resolves: null }],
      errors: [
        {
          messageId: 'restrictedChain',
          data: {
            message: null,
            restriction: 'resolves',
          },
          column: 11,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(a).resolves.not.toBe(b)',
      options: [{ resolves: null }],
      errors: [
        {
          messageId: 'restrictedChain',
          data: {
            message: null,
            restriction: 'resolves',
          },
          column: 11,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(a).resolves.not.toBe(b)',
      options: [{ 'resolves.not': null }],
      errors: [
        {
          messageId: 'restrictedChain',
          data: {
            message: null,
            restriction: 'resolves.not',
          },
          column: 11,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(a).not.toBe(b)',
      options: [{ 'not.toBe': null }],
      errors: [
        {
          messageId: 'restrictedChain',
          data: {
            message: null,
            restriction: 'not.toBe',
          },
          endColumn: 19,
          column: 11,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(a).resolves.not.toBe(b)',
      options: [{ 'resolves.not.toBe': null }],
      errors: [
        {
          messageId: 'restrictedChain',
          data: {
            message: null,
            restriction: 'resolves.not.toBe',
          },
          endColumn: 28,
          column: 11,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(a).toBe(b)',
      options: [{ toBe: 'Prefer `toStrictEqual` instead' }],
      errors: [
        {
          messageId: 'restrictedChainWithMessage',
          data: {
            message: 'Prefer `toStrictEqual` instead',
            restriction: 'toBe',
          },
          column: 11,
          line: 1,
        },
      ],
    },
    {
      code: `
        test('some test', async () => {
          await expect(Promise.resolve(1)).resolves.toBe(1);
         });
      `,
      options: [{ resolves: 'Use `expect(await promise)` instead.' }],
      errors: [
        {
          messageId: 'restrictedChainWithMessage',
          data: {
            message: 'Use `expect(await promise)` instead.',
            restriction: 'resolves',
          },
          endColumn: 57,
          column: 44,
        },
      ],
    },
    {
      code: 'expect(Promise.resolve({})).rejects.toBeFalsy()',
      options: [{ 'rejects.toBeFalsy': null }],
      errors: [
        {
          messageId: 'restrictedChain',
          data: {
            message: null,
            restriction: 'rejects.toBeFalsy',
          },
          endColumn: 46,
          column: 29,
        },
      ],
    },
    {
      code: "expect(uploadFileMock).not.toHaveBeenCalledWith('file.name')",
      options: [
        { 'not.toHaveBeenCalledWith': 'Use not.toHaveBeenCalled instead' },
      ],
      errors: [
        {
          messageId: 'restrictedChainWithMessage',
          data: {
            message: 'Use not.toHaveBeenCalled instead',
            restriction: 'not.toHaveBeenCalledWith',
          },
          endColumn: 48,
          column: 24,
        },
      ],
    },
  ],
});

File: internal/plugins/jest/rules/no_restricted_matchers/no_restricted_matchers_test.go
package no_restricted_matchers_test

import (
	"testing"

	"github.com/web-infra-dev/rslint/internal/plugins/jest/fixtures"
	"github.com/web-infra-dev/rslint/internal/plugins/jest/rules/no_restricted_matchers"
	"github.com/web-infra-dev/rslint/internal/rule_tester"
)

func TestNoRestrictedMatchersRule(t *testing.T) {
	rule_tester.RunRuleTester(
		fixtures.GetRootDir(),
		"tsconfig.json",
		t,
		&no_restricted_matchers.NoRestrictedMatchersRule,
		[]rule_tester.ValidTestCase{
			{Code: `expect(a).toHaveBeenCalled()`},
			{Code: `expect(a).not.toHaveBeenCalled()`},
			{Code: `expect(a).toHaveBeenCalledTimes()`},
			{Code: `expect(a).toHaveBeenCalledWith()`},
			{Code: `expect(a).toHaveBeenLastCalledWith()`},
			{Code: `expect(a).toHaveBeenNthCalledWith()`},
			{Code: `expect(a).toHaveReturned()`},
			{Code: `expect(a).toHaveReturnedTimes()`},
			{Code: `expect(a).toHaveReturnedWith()`},
			{Code: `expect(a).toHaveLastReturnedWith()`},
			{Code: `expect(a).toHaveNthReturnedWith()`},
			{Code: `expect(a).toThrow()`},
			{Code: `expect(a).rejects;`},
			{Code: `expect(a);`},
			{
				Code: `expect(a).resolves`,
				Options: []interface{}{
					map[string]interface{}{"not": nil},
				},
			},
			{
				Code: `expect(a).toBe(b)`,
				Options: []interface{}{
					map[string]interface{}{"not.toBe": nil},
				},
			},
			{
				Code: `expect(a).toBeUndefined(b)`,
				Options: []interface{}{
					map[string]interface{}{"toBe": nil},
				},
			},
			{
				Code: `expect(a)["toBe"](b)`,
				Options: []interface{}{
					map[string]interface{}{"not.toBe": nil},
				},
			},
			{
				Code: `expect(a).resolves.not.toBe(b)`,
				Options: []interface{}{
					map[string]interface{}{"not": nil},
				},
			},
			{
				Code: `expect(a).resolves.not.toBe(b)`,
				Options: []interface{}{
					map[string]interface{}{"not.toBe": nil},
				},
			},
			{
				Code: `expect(uploadFileMock).resolves.toHaveBeenCalledWith('file.name')`,
				Options: []interface{}{
					map[string]interface{}{"not.toHaveBeenCalledWith": "Use not.toHaveBeenCalled instead"},
				},
			},
			{
				Code: `expect(uploadFileMock).resolves.not.toHaveBeenCalledWith('file.name')`,
				Options: []interface{}{
					map[string]interface{}{"not.toHaveBeenCalledWith": "Use not.toHaveBeenCalled instead"},
				},
			},
		},
		[]rule_tester.InvalidTestCase{
			{
				Code: `expect(a).toBe(b)`,
				Options: []interface{}{
					map[string]interface{}{"toBe": nil},
				},
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "restrictedChain",
						Message:   "Use of `toBe` is disallowed",
						Line:      1,
						Column:    11,
					},
				},
			},
			{
				Code: `expect(a)["toBe"](b)`,
				Options: []interface{}{
					map[string]interface{}{"toBe": nil},
				},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "restrictedChain", Line: 1, Column: 11},
				},
			},
			{
				Code: `expect(a).not[x]()`,
				Options: []interface{}{
					map[string]interface{}{"not": nil},
				},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "restrictedChain", Line: 1, Column: 11},
				},
			},
			{
				Code: `expect(a).not.toBe(b)`,
				Options: []interface{}{
					map[string]interface{}{"not": nil},
				},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "restrictedChain", Line: 1, Column: 11},
				},
			},
			{
				Code: `expect(a).resolves.toBe(b)`,
				Options: []interface{}{
					map[string]interface{}{"resolves": nil},
				},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "restrictedChain", Line: 1, Column: 11},
				},
			},
			{
				Code: `expect(a).resolves.not.toBe(b)`,
				Options: []interface{}{
					map[string]interface{}{"resolves": nil},
				},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "restrictedChain", Line: 1, Column: 11},
				},
			},
			{
				Code: `expect(a).resolves.not.toBe(b)`,
				Options: []interface{}{
					map[string]interface{}{"resolves.not": nil},
				},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "restrictedChain", Line: 1, Column: 11},
				},
			},
			{
				Code: `expect(a).not.toBe(b)`,
				Options: []interface{}{
					map[string]interface{}{"not.toBe": nil},
				},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "restrictedChain", Line: 1, Column: 11, EndColumn: 19},
				},
			},
			{
				Code: `expect(a).resolves.not.toBe(b)`,
				Options: []interface{}{
					map[string]interface{}{"resolves.not.toBe": nil},
				},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "restrictedChain", Line: 1, Column: 11, EndColumn: 28},
				},
			},
			{
				Code: `expect(a).resolves.not.toBe(b)`,
				Options: []interface{}{
					map[string]interface{}{
						"resolves":     "Do not use resolves",
						"resolves.not": "Do not use resolves.not",
					},
				},
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "restrictedChainWithMessage",
						Message:   "Do not use resolves.not",
						Line:      1,
						Column:    11,
						EndColumn: 28,
					},
				},
			},
			{
				Code: `expect(a).toBe(b)`,
				Options: []interface{}{
					map[string]interface{}{"toBe": "Prefer `toStrictEqual` instead"},
				},
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "restrictedChainWithMessage",
						Message:   "Prefer `toStrictEqual` instead",
						Line:      1,
						Column:    11,
					},
				},
			},
			{
				Code: `
        test('some test', async () => {
          await expect(Promise.resolve(1)).resolves.toBe(1);
         });
      `,
				Options: []interface{}{
					map[string]interface{}{"resolves": "Use `expect(await promise)` instead."},
				},
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "restrictedChainWithMessage",
						Message:   "Use `expect(await promise)` instead.",
						Column:    44,
						EndColumn: 57,
					},
				},
			},
			{
				Code: `expect(Promise.resolve({})).rejects.toBeFalsy()`,
				Options: []interface{}{
					map[string]interface{}{"rejects.toBeFalsy": nil},
				},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "restrictedChain", Column: 29, EndColumn: 46},
				},
			},
			{
				Code: `expect(uploadFileMock).not.toHaveBeenCalledWith('file.name')`,
				Options: []interface{}{
					map[string]interface{}{"not.toHaveBeenCalledWith": "Use not.toHaveBeenCalled instead"},
				},
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "restrictedChainWithMessage",
						Message:   "Use not.toHaveBeenCalled instead",
						Column:    24,
						EndColumn: 48,
					},
				},
			},
		},
	)
}
